File: python/euler/bin/p0035.py
# project euler: problem 35

from euler.lib.prime import is_prime, primes
from euler.lib.util import num_of_digits


def compute(limit: int) -> str:
    def is_circular_prime(n: int) -> bool:
        k = num_of_digits(n) - 1
        d = pow(10, k)
        for _ in range(k):
            n = (n % 10) * d + n // 10
            if is_prime(n) is False:
                return False

        return True

    return str(len([n for n in primes(limit) if is_circular_prime(n) is True]))


def solve() -> str:
    return compute(1_000_000)

File: typescript/euler/bin/p0035.ts
// project euler: problem 35

import { isPrime, primes } from "../lib/prime.ts";
import { range } from "../lib/util.ts";

export const compute = (limit: number): string => {
  const checkRotNum = (n: number): boolean => {
    const s = String(n) + String(n);
    const m = Math.trunc(s.length / 2);

    for (const pos of range(0, m)) {
      if (!isPrime(Number(s.slice(pos, pos + m)))) {
        return false;
      }
    }

    return true;
  };

  let acc = 0;
  for (const n of primes(limit)) {
    if (checkRotNum(n)) {
      acc += 1;
    }
  }

  return String(acc);
};

export const solve = (): string => compute(1_000_000);
